app/forms/ip-pool-range-add.tsx

```diff
@@ -0,0 +1,111 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, you can obtain one at https://mozilla.org/MPL/2.0/.
+ *
+ * Copyright Oxide Computer Company
+ */
+import { useNavigate } from 'react-router-dom'
+
+import { useApiMutation, useApiQueryClient, type IpRange } from '@oxide/api'
+import { Message } from '@oxide/ui'
+import { IPV4_REGEX, IPV6_REGEX } from '@oxide/util'
+
+import { SideModalForm, TextField } from 'app/components/form'
+import { useForm, useIpPoolSelector } from 'app/hooks'
+import { addToast } from 'app/stores/toast'
+import { pb } from 'app/util/path-builder'
+
+const defaultValues: IpRange = {
+  first: '',
+  last: '',
+}
+
+function validateIp(s: string) {
+  const isv4 = IPV4_REGEX.test(s)
+  const isv6 = !isv4 && IPV6_REGEX.test(s)
+  return { isv4, isv6, valid: isv4 || isv6 }
+}
+
+const invalidAddressError = { type: 'pattern', message: 'Not a valid IP address' }
+
+const diffVersionError = {
+  type: 'custom',
+  message: 'First and last must be the same version',
+}
+
+/**
+ * Pretty straightforward -- make sure IPs are valid and both first and last
+ * are the same version. Because we're putting the same version error on both
+ * fields, this could also work as a regular validate() on each field, where
+ * each field compares itself to the other. It seems silly to run the giant
+ * regex twice, though.
+ */
+function resolver(values: IpRange) {
+  const first = validateIp(values.first)
+  const last = validateIp(values.last)
+
+  let errors = undefined
+
+  if (!first.valid || !last.valid) {
+    errors = {
+      first: first.valid ? undefined : invalidAddressError,
+      last: last.valid ? undefined : invalidAddressError,
+    }
+  } else if ((first.isv4 && last.isv6) || (first.isv6 && last.isv4)) {
+    errors = { first: diffVersionError, last: diffVersionError }
+  }
+
+  // TODO: if we were really cool we could check first <= last but that sounds
+  // like a pain
+
+  return errors ? { values: {}, errors } : { values, errors: {} }
+}
```

Comment:
It turns out writing a custom resolver to validate multiple fields together is very easy. We could do this more often. The only problem is that I think it can't be used with other kinds of validation. For example, I think the `required` prop below is being ignored because of this resolver. In this case, that's fine because empty string is not valid IP, so we catch it here.